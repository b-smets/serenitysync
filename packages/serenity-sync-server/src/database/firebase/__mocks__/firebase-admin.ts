import { MockFirebase, MockFirebaseSdk } from 'firebase-mock';

interface MockFirebaseSdkExt extends MockFirebaseSdk {
  credential: {
    cert: () => void;
  };
  __setMockData: (data: any) => void;
  __clear: () => void;
}

const createDatabase = (data: any) => {
  const newDatabase = new MockFirebase(undefined, data);
  newDatabase.autoFlush(true);
  return newDatabase;
};

let database: MockFirebase = createDatabase(undefined);

const firebase = new MockFirebaseSdk(() => database) as MockFirebaseSdkExt;
firebase.credential = {
  cert: jest.fn(),
};
firebase.__setMockData = (data: any) => {
  database = createDatabase(data);
};
firebase.__clear = () => {
  database = createDatabase(undefined);
};

export = firebase;
