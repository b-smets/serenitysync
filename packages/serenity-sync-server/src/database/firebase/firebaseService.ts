import * as firebase from 'firebase-admin';
import { calculateRevisionID } from '../../util/revisions';
import { DocumentResult } from '../../schema';

const serviceAccount = require('../../../config/firebase.json');

const app = firebase.initializeApp({
  credential: firebase.credential.cert(serviceAccount),
  databaseURL: 'https://sync-1b17c.firebaseio.com',
});

const database = app.database();

interface Document {
  id: string;
  rev: string;
  body: any;
}

const prepareEntryForTransport = (entry: Document) => ({ ...entry, body: JSON.stringify(entry.body) });

export const documents = (): Promise<DocumentResult[]> =>
  database.ref('documents').once('value')
    .then((snap: firebase.database.DataSnapshot) => {
      if (snap.exists()) {
        const value = snap.val();
        return Object.keys(value).map(key => prepareEntryForTransport(value[key]));
      }
      return [];
    });

export const documentByID = (id: string): Promise<DocumentResult> =>
  database.ref(`documents/${id}`).once('value')
    .then((snap: firebase.database.DataSnapshot) => {
      if (snap.exists()) {
        return prepareEntryForTransport(snap.val());
      }
      throw new Error('Not found');
    });

export const addDocument = (id: string, body: string, rev?: string): Promise<DocumentResult> => {
  const ref = database.ref(`documents/${id}`);
  return ref.once('value')
    .then((snap: firebase.database.DataSnapshot) => {
      if (snap.exists()) {
        const current = snap.val();
        if (current.rev !== rev) {
          throw new Error('Revision conflict: ' + current.rev);
        }
      }
      const newRevSequenceNumber = calculateRevisionID(body, rev);
      const entry = {
        id,
        rev: newRevSequenceNumber,
        body: JSON.parse(body),
      };
      return ref.set(entry).then(() => ({ ...entry, body }));
    });
};
