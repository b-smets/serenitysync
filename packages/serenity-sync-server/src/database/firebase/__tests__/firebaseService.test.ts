jest.unmock('../firebaseService');

import * as firebase from 'firebase-admin';
import { documents } from '../firebaseService';

describe('Firebase Service', () => {

  describe('documents', () => {

    beforeEach(() => {
      (firebase as any).__clear();
    });

    test('obtains a list of documents from the firebase', () => {
      (firebase as any).__setMockData({
        foo: {
          id: 'foo',
          rev: '1-abc',
          body: {
            one: 'alpha',
            two: 'beta',
            three: 'gamma',
          },
        },
        bar: {
          id: 'bar',
          rev: '2-def',
          body: {
            three: 'delta',
            four: 'epsilon',
            five: 'omega',
          }
        }
      });
      return documents('myDB').then(result => {
        expect(result).toMatchSnapshot();
      });
    });

    test('obtains an empty list if the database is empty', () => {
      (firebase as any).__setMockData({});
      return documents('myDB').then(result => {
        expect(result).toEqual([]);
      });
    });

    test('obtains an empty list if the database is unknown', () => {
      return documents('myDB').then(result => {
        expect(result).toEqual([]);
      });
    });

    test('rejects if retrieving the data fails', () => {
      const expectedError = new Error('fail');
      (firebase.database().ref() as any).failNext('once', expectedError);
      return documents('myDB').then(result => fail('should have been rejected'))
        .catch(error => expect(error).toBe(expectedError));
    });
  });

  // describe('documentByID', () => {

  // });

  // describe('addDocument', () => {

  // });

  // describe('deleteDocument', () => {

  // });

});
