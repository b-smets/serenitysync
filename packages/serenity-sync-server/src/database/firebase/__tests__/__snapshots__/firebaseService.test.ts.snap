// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Firebase Service documents obtains a list of documents from the firebase 1`] = `
Array [
  Object {
    "body": "{\\"three\\":\\"delta\\",\\"four\\":\\"epsilon\\",\\"five\\":\\"omega\\"}",
    "id": "bar",
    "rev": "2-def",
  },
  Object {
    "body": "{\\"one\\":\\"alpha\\",\\"two\\":\\"beta\\",\\"three\\":\\"gamma\\"}",
    "id": "foo",
    "rev": "1-abc",
  },
]
`;
